---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# tapestri.tools

<!-- badges: start -->

<!-- badges: end -->

A toolbox for the analysis of MissionBio's tapestri single cell multi-omics data.

## Installation

You can install the development version of tapestri.tools from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("northNomad/tapestri.tools")
```

## Load tapestri multi-omics data stored in .h5 format into session

```{r}
library(tapestri.tools)

h5f <- tapestri.tools::read_h5(path = "data_private/mpdxD-18_MB_33_38.dna+protein.h5")
```

## Read all the variants called by haplotype caller

```{r}
dt_variants <- get_variants(h5f = h5f, format = "data.table")
head(dt_variants)
```

## `tapestri.tools` makes data wrangling easy

In the example experiment, we transplanted a mix of two primary AML samples into one immunodeficient mice.

12 weeks after transplantation, we harvested the bone marrow cells and performed single cell DNA + protein sequencing using MissionBio's tapestri platform.

One of the AML sample carries IDH1^R132H^ mutation, the other carries IDH2^R140Q^; we know these two *IDH* variants are mutually exclusive.

```{r}
#based on hg19
int_variants <- c(IDH1_R132H = "chr2:209113112:C/T", IDH2_R140Q = "chr15:90631934:C/T")

tapestri.tools::plot_upset_with_variants(h5f, variants = int_variants)
```


### Filter or index the cells based on the genotypes of specific variants:

```{r}
#which cells carry IDH1 mutation? 
index_cells_IDH1 <- get_cells_index(h5f, variants = int_variants[1], int_ngt = c(1, 2)) 

#which cells carry IDH2 mutation?
index_cells_IDH2 <- get_cells_index(h5f, variants = int_variants[2], int_ngt = c(1, 2))

#which cells are doublets 
x <- get_cells_index(h5f, variants = int_variants, int_ngt = c(1, 2)) 
intersect(x$IDH1_R132H, x$IDH2_R140Q)
```


### Read in genotype matrix of FLT3ITD variants of IDH1 mutated cells

```{r}
id_FLT3ITD <- get_flt3itd(h5f)$id
names(id_FLT3ITD) <- paste0("FLT3ITD", 1:length(id_FLT3ITD))

read_assays_variants(h5f, 
                     included_assays = c("AF", "NGT"),
                     index_variants = get_variants_index(h5f, id_FLT3ITD),
                     index_cells = index_cells_IDH1,  
                     format = "list") -> x

names(x) #AF, NGT
dim(x$AF)
```


....More plotting and filtering functions to come...

What is special about using `README.Rmd` instead of just `README.md`? You can include R chunks like so:

```{r cars}
summary(cars)
```

You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this. You could also use GitHub Actions to re-render `README.Rmd` every time you push. An example workflow can be found here: <https://github.com/r-lib/actions/tree/v1/examples>.

You can also embed plots, for example:

```{r pressure, echo = FALSE}
plot(pressure)
```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub and CRAN.
